### DISCUSSION TITLE
What kinds of workflows do users have?

### DISCUSSION THREAD

**Discussion by acmcarther** [google/cargo-raze#327] [Ideas]
### Summary of the discussion as of DEC 23 AM

We introduce two terms for mental models of how users might use cargo-raze. They are:
- "tomls-everywhere" model: Users define much or all of their crates through Cargo.tomls, including internal ones.
- "no-internal-tomls" model: Users define their internal crates (libraries, binaries, etc) using regular rules_rust and BUILD files, and use one or a few Cargo.tomls to define the full set of third_party dependencies, or some new invention like a `crate_universe` rule.

(this isn't necessarily meant to say one could not apply both approached, but that they're distinct ways of thinking about the cargo-bazel integration)

@UebelAndre and I agree that both mental models (which map down to code organization strategies) have valid usecases and should be supported. I add  here now that we should view all new features as foundational (meaning internal or applies to both), or as applying to support one or the other usecase. I further suggest that we include a section very high in the documentation describing the two models, and helping users decide which works best for their Bazel workspace(s).

In general, the rest of the discussion consists of how we  developed the previous vocabulary, and details in particular about the future direction of the "tomls-everywhere" model.

### Original Post Below

---

Branched from discussion in https://github.com/google/cargo-raze/pull/319

The workflow type that I'm familiar with (cribbed from how Google manages third party dependencies) is that there is a single repository of third party crates in the monorepo. In general, there is a rule there that there be only one version of each dependency. For each language, a set of special arbiters of third party dependencies exists. These folks gatekeep the inclusion of new third party code.

In that model, the organization in general would only have one cargo.toml, because the organization would not integrate deeply with cargo. That singular cargo.toml would not map to any single application. Instead application developers write applications and libraries more like one would write C++ applications and libraries, using the blessed versions of dependencies in `//third_party` or whatever, and never interact with cargo.toml.

The repo structure ends up looking like this

```
/third_party/crates/Cargo.toml
/third_party/crates/some_crate/
/some_sub_organization/some_project/some_component/code.rs 
```
---

In #319, @UebelAndre alludes to I think a different workflow which is more deeply integrated into cargo, and in general treats the cargo.toml, not the BUILD file, as the source of truth. I think this is an interesting model that should not necessarily be dismissed, in large part because it's significantly closer to the existing cargo user experience.

I think that workspace looks more like this

```
/some_sub_organization/some_project/Cargo.toml
/some_sub_organization/some_project/sub_crate/code.rs
/some_sub_organization/some_project/other_sub_crate/code.rs
/some_sub_organization/some_project/dependencies/some_crate
/some_sub_organization/some_other_project/Cargo.toml
/some_sub_organization/some_other_project/sub_crate/code.rs
/some_sub_organization/some_other_project/other_sub_crate/code.rs
/some_sub_organization/some_other_project/dependencies/some_crate
```

**Comment by acmcarther** [comment]
In general, I don't think I want to hash out here which model is objectively better. The google approach is [detailed here](https://opensource.google/docs/thirdparty/oneversion/). I think we can probably agree that (1) that approach might be better for large organizations, and (2) that approach might be onerous for small organizations. Furthermore, I think we can agree that the google approach forces code organization decisions that break assumptions made by most existing rust tools, which is a huge bummer.

---

In https://github.com/google/cargo-raze/pull/319#issuecomment-749806340, @UebelAndre says

> We have a few mono-ish repos that would immediately make use of cargo workspaces. We want to be able to take advantage of the lockfiles and avoid having a single volatile/hand authored place for all dependencies. And we don't want to have to be spending all our time writing toml and BUILD files. In an ideal world, we would near seamlessly replace the use of Cargo with Bazel and be able to do all the same things without a bunch of additional hassle.

It sounds like you envision developers on individual projects writing cargo.toml files for each crate, but instead of building with cargo, you'd build with bazel. Is that accurate? Under what circumstances would developers on the same or different project or same or different organization try to ensure that the same versions of dependencies are used?

**Reply by UebelAndre** [reply]
Cargo has tooling for that that `cargo-raze` takes advantage of, the lockfile. It's otherwise not a requirement that every crate in a repo use the exact same version of the dependencies. If one crate wants to use `tokio=0.3` and another wants to use `tokio=0.2` then that should be allowed, especially if the two crates are not directly related to each other (tools or libraries for different components of a larger project).

**Reply by acmcarther** [reply]
I think in the first philosophy (the one-version-rule), the organization would mandate that in-organization code all use the same version of a given dependency. While third party dependencies may depend on multiple versions of the same dependency, this would just be a pragmatic concession, not a deliberate choice. That reality need not be something set in stone either. In the Haskell ecosystem, Cabal('s dependency system) is analogous to Cargo here, and I think the Rust ecosystem could eventually see an analogue to [Stackage](https://www.stackage.org/), which is basically a curated set of packages which all interoperate without the need for multiple dependency versions.

**Reply by UebelAndre** [reply]
Not every organization is equipped to assert that organization wide, specific versions of dependencies will be used. I don't think it'd be appropriate for this tool to try and enforce that. Both cases should be supported.

**Reply by acmcarther** [reply]
Hence, why this topic exists.

**Reply by UebelAndre** [reply]
So, the only two workflows I personally keep in mind is:
1. What is the existing workflow for using standard tools of a particular language
2. What is the most Bazel-ish thing to do to keep everything fast and correct.

I think it's valuable to support the use case that someone has a cargo project and wants to leave it as is but add Bazel files to start building with Bazel.

Additionally, there are some freedoms you can take if you're *only* going to be using Bazel and have never had cargo in your project. I think this project has exclusively lived in this camp for the majority of it's life but being new to Rust but familiar with Bazel, I found it more natural and easier to be able to keep both Cargo and Bazel working side by side so I can prototype something new and weird in the environment where it was designed and then begin to port it over to Bazel. Without that, many interactions around some `build.rs` file or use of other crates that rely on Cargo metadata ([cbindgen](https://github.com/eqrion/cbindgen)) might have driven me to simply abandon Bazel or come up with my own rules backed by cargo (similar to what happened in https://github.com/google/cargo-raze/pull/193).

**Reply by acmcarther** [reply]
Couple of questions

- How should that user start using Bazel for their existing rust crate? Do they clone it into an existing monorepo? Where is it situated if there are preexisting crates? How does a third library depend upon both crates?
- Whats the expected level of backward and forward compatibility? A crate using rust_proto_library rules and then depending upon them in a rust library would not be backward compatible with cargo

**Reply by UebelAndre** [reply]
> How should that user start using Bazel for their existing rust crate? 

In an ideal world, `cargo-raze` could setup their Cargo project to start using Bazel similar to [wasm-pack](https://github.com/rustwasm/wasm-pack) or [`cargo init`](https://doc.rust-lang.org/cargo/commands/cargo-init.html). 

> Do they clone it into an existing monorepo? Where is it situated if there are preexisting crates? 

The crates can exist in a mono repo or a not. The repo or location is not important, so long as it has a valid cargo workspace (purely speaking in terms structure), raze can generate the necessary files to support it in Bazel. A third 

> How does a third library depend upon both crates?

A third library could be added either in Bazel or Cargo.

> Whats the expected level of backward and forward compatibility? A crate using rust_proto_library rules and then depending upon them in a rust library would not be backward compatible with cargo

If users decide to use other Bazel targets in as dependencies for their crates, then it should not be backwards compatible. Cargo would only be able to represent a dependency tree moving forward.

**Comment by UebelAndre** [comment]
I'll add more context here but:

> In #319, @UebelAndre alludes to I think a different workflow which is more deeply integrated into cargo, and in general treats the cargo.toml, not the BUILD file as the source of truth

This is currently how workflows involving `cargo-raze` work the source of truth is not actually what's written in the BUILD files, it's whatever `cargo-raze` generated and aliased in the `/third_party/crates` package. In an ideal future, there would be a way to author `rust_*` target in a BUILD file and have Bazel correctly fetch the intended dependencies, but no such tooling exists today. The workflow I'm interested in is taking existing Cargo projects and converting them to use Bazel. The only future dependency on cargo is to define dependencies in a way that `cargo-raze` could generate a way for Bazel to fetch them.

**Reply by acmcarther** [reply]
Cargo.toml in the first model is the source of truth only the list of third party dependencies (and the BUILD rules for building them), but it is not authoritative for the in-organization code, which is built using regular BUILD rules in regular BUILD files. 

Perhaps we should talk more about what this `have Bazel correctly fetch the intended dependencies` would look like, because I think this is a component of the decentralized more-cargo-esque philosophy.

**Reply by UebelAndre** [reply]
The fundamental driving factor behind https://github.com/google/cargo-raze/pull/319 is that in order to build crates that depend on other crates hosted on some registry somewhere is that I don't want to have to specify dependencies multiple times. Right now, even if you define crates all in one `Cargo.toml` file, you have to add it there and then go add it to your BUILD file(s). How can that be solved for? My thought is that the Cargo project structure could be used to allow `cargo-raze` to be able to solve this problem by generating macros for each package. I think very few companies have gigantic repos like google does where virtually every crate you might ever need is already there and being used.

**Reply by acmcarther** [reply]
This "i am duplicating my dependencies" problem seems like a result of the desire to use Cargo.tomls to as the source of truth for non-third-party code, which you might interpret as "the natural approach" (particularly given the example of migrating an existing crate), but I haven't historically treated this as the main objective.

Below we've called this the "declare-in-cargo-build-in-bazel" model (which is already tiresome and needs a new name). I believe this is your prevailing view of how the tool should work, contrary to my "i-hate-cargo-bazel-only-please" model. 

I decree these models henceforth be called
- tomls-everywhere-model: All rust third-party and internal is defined through Cargo.toml files. BUILD files are simply generated artifacts derived from the tomls.
- no-internal-tomls-model: Third party code is (by lack of an alternative) defined through cargo tomls. Internal code uses BUILD files
(at least, through the end of this post)

My primary concern, which I somehow expend huge numbers of words yet fail to communicate, is that I don't understand how the "tomls-everywhere-model" deals with features that Bazel supports but Cargo does not. The no-internal-tomls model deals with this by ditching the tomls at the boundary.

**Reply by UebelAndre** [reply]
> This "i am duplicating my dependencies" problem seems like a result of the desire to use Cargo.tomls to as the source of truth for non-third-party code

This is not quite true. Even in `no-internal-tomls-model` if I have to go write my dependency in some `cargo_universe` definition, and *additionally* plug it into my dependencies, I'll have the exact same frustration. 

> My primary concern, which I somehow expend huge numbers of words yet fail to communicate, is that I don't understand how the "tomls-everywhere-model" deals with features that Bazel supports but Cargo does not. The no-internal-tomls model deals with this by ditching the tomls at the boundary.

In `tomls-everywhere-model`, users are making a choice to not use the Bazel features you're referring to. In the case where you're simply tying to add Bazel support to a cargo project, this is fine (referring to the `wasm-bindgen-cli` setup).

I may have been unclear before when I talked about Cargo/Bazel interchangeability. I want to make the transition between Cargo and Bazel smooth and I want to have a tool I can use to add `new_*` workspace dependencies to my Bazel project and generate Bazel files to build the Cargo project in said dependency.

**Reply by acmcarther** [reply]
> This is not quite true. Even in no-internal-tomls-model if I have to go write my dependency in some cargo_universe definition, and additionally plug it into my dependencies, I'll have the exact same frustration.

This assumes one of two things i think
a. You use the same set of dependencies for all non-third-party code (maybe?) and have multiple packages for a given cargo_universe rule, and hence are burdened with duplicating the full dependency list every time even though it is redundant
b. You have a one to one ratio of non-third-party packages to cargo_universe rules, and hence, always use the same dependencies that you specify in the toml as those you specify in the BUILD rule

If not (a), then the dependencies are not redundant. If not (b), then its unlikely that the many sets of targets across the bazel workspace all share the same dependency set.

> I may have been unclear before when I talked about Cargo/Bazel interchangeability. I want to make the transition between Cargo and Bazel smooth and I want to have a tool I can use to add new_* workspace dependencies to my Bazel project and generate Bazel files to build the Cargo project in said dependency.

I'm actually still not really clear what this means. I think this means that you'd like to be able to bring in an arbitrary new_git_repository and have it hook into the workspace's existing third party crates?

**Reply by UebelAndre** [reply]
> If not (a), then the dependencies are not redundant. If not (b), then its unlikely that the many sets of targets across the bazel workspace all share the same dependency set.

Steps to add a dependency:
1. Go add it to your `Cargo.toml` file
2. run `cargo raze`
3. Add the alias it spits out to your Bazel target

Step 3 should be eliminated. 

I fully plan to have a 1-1 mapping of `crate_universe` to package. I do not have the luxury of being able to dictate that every single target use the same versions of dependencies.

Naturally, if I were to have a single `cargo_unvierse` for all my dependencies, then it's likely the case step 1 and 2 almost never happen and step 3 is the only thing I'd need to do.

> I'm actually still not really clear what this means. I think this means that you'd like to be able to bring in an arbitrary new_git_repository and have it hook into the workspace's existing third party crates?

I added `binary_deps` for the sole purpose of bringing back missing functionality to support this:
https://github.com/bazelbuild/rules_rust/blob/c7a408a8c58c6dde73a552d6f78f90c7ef96b513/wasm_bindgen/raze/Cargo.toml#L17-L18

But since implementing it, I feel it was the wrong choice. The issue is that `wasm-bindgen-cli` is a binary only crate that has a dependency on it's larger library. Because the crate does not have a library, it will not show up in `cargo metadata`. So the question remains, how do I build the `wasm-bindgen-cli` binary to use in the `rust_wasm_bindgen` rules?

The solution I'm trying to get to is: Bring in `wasm-bindgen-cli` as a `http_archive` and provide a dedicated BUILD file for it. This is ultimately exactly what's being done already:
https://github.com/bazelbuild/rules_rust/blob/c7a408a8c58c6dde73a552d6f78f90c7ef96b513/wasm_bindgen/raze/crates.bzl#L1544-L1552

I want to create a `razed_crate_repository` (or something) repository rule that I can instead write
```python
razed_crate_repository(
    name = "wasm_bindgen_cli",
    crate = "wasm-bindgen-cli",
    version = "0.2.69",
)
```

and be able to build the binary via `bazel build @wasm_bindgen_cli//:wasm_bindgen_cli`. So in order for that to work, I need something that works with `tomls-everywhere-model` because that's exclusively the model this rule will use.

**Reply by acmcarther** [reply]
> I fully plan to have a 1-1 mapping of crate_universe to package. I do not have the luxury of being able to dictate that every single target use the same versions of dependencies.

I don't intend to block future work on this question, but how do you intend to make crates which have independent crate_universes interoperate? Are they permitted to have dependency relationships between one another (or is another crate permitted to depend upon both of them)?

> I want to create a razed_crate_repository (or something) repository rule that I can instead write (...)

Ah, I remember this idea from the ancient times. It was once called `cargo_crate`. To start with, you'll want that build rule to accept a dependency repository (instead of grabbing things from the "opaque universe", probably).

**Reply by UebelAndre** [reply]
> I don't intend to block future work on this question, but how do you intend to make crates which have independent crate_universes interoperate? Are they permitted to have dependency relationships between one another (or is another crate permitted to depend upon both of them)?

I think that's a fantastic question to go on https://github.com/bazelbuild/rules_rust/issues/2 I would expect two crates building from separate `cargo_universe`s to build without any shared dependencies. If users want to have their dependencies shared, then they should subscribe to the `no-internal-toml-model` and use a single universe. Maybe there's some kind of `cargo workspace` abstraction that can be done to sorta streamline that. 

> Ah, I remember this idea from the ancient times. It was once called `cargo_crate`. To start with, you'll want that build rule to accept a dependency repository (instead of grabbing things from the "opaque universe", probably).

What do you mean "dependency repository"? 

My plan so far was to have the package built and published somewhere by Bazel such that a repository rule could download and execute a binary to render what it needed into the new repository. Then users would load some rendered `bzl` file to add all the transitive dependencies to the workspace. I find this really exciting because I feel there's some incarnation of this that might even be able to help users with building complex crates like things that require `curl`, `openssl`, `zlib`, `libssh` (looking at you `cargo-raze` 😆) which are otherwise pretty tricky to build.

**Reply by acmcarther** [reply]
> I think that's a fantastic question to go on bazelbuild/rules_rust#2 I would expect two crates building from separate cargo_universes to build without any shared dependencies. If users want to have their dependencies shared, then they should subscribe to the no-internal-toml-model and use a single universe. Maybe there's some kind of cargo workspace abstraction that can be done to sorta streamline that.

There is a perhaps not intractable problem at the rustc level here which has dissuaded me from this in the past:

An example:
- crate_universe_1
- crate_universe_2
- my_crate1 (depends on crate_universe1)
- my_crate2 (depends on crate_universe2)
- my_crate3 (depends on my_crate1 and my_crate2)

Compilation of my_crate3 maps down into a rust_library or rust_binary which may encounter the two of the same dependency@version which was configured with different options in each crate universe. In vanilla cargo, the options are reconciled during dependency resolution (resulting in both compilation flags, and potential new or optional dependencies), and in the process of reconciliation the two crates depending on dependency@version are stuck with the same crate with the same features.

EDIT: this duplication is bad because the eventual rustc invocation for my_crate3 will be very unhappy to see the same rlib.

EDIT: the model here prevents the same type of reconciliation as vanilla cargo, because dependencies and their flags must be resolved at crate_universe declaration time. There's probably a metaphor I could make here to what vanilla cargo does... maybe think of what you're trying here as "partially compiling two cargo crates via two separate invocations, then trying to join them using a third"

> What do you mean "dependency repository"?

Basically, in order to automagically the whole `razed_crate_repository('serde')`, you need to have a single global crate universe which is brought in by the build rule itself, or is accessed via some `bind` stuff. Or you can pass in what i call a "dependency registry" or "dependency repository" so that you can have more than one.

**Reply by UebelAndre** [reply]
> EDIT: the model here prevents the same type of reconciliation as vanilla cargo, because dependencies and their flags must be resolved at crate_universe declaration time. There's probably a metaphor I could make here to what vanilla cargo does... maybe think of what you're trying here as "partially compiling two cargo crates via two separate invocations, then trying to join them using a third"

I'm not too worried about this issue. First, my thoughts for `razed_crate_repository` is that it should be used in cases where you need a stand alone binary (maybe library too but in some weird case).

I'm not sure if it's clear but a `cargo_universe` to me equates to a cargo workspace. So it can contain effectively multiple toml files but everything is shared in the same pool. The benefit to using a `cargo_universe` that is built by multiple toml files from a cargo workspace is that associated dependencies can be reasoned about by paring the cargo package path to the bazel package path. Which is what I'm doing in https://github.com/google/cargo-raze/pull/319

I can see what you're saying about that third crate though. Good to know 🙏

**Reply by acmcarther** [reply]
I sometimes cant tell if you understand an underlying issue more completely or less completely than I do, and we end up in cycles of explaining the same (or sometimes different) things to one another.

I interpret `cargo_universe` not as a cargo workspace, but rather as a pass of "cargo dependency and feature resolution"  over a root crate's dependencies and features. This is effectively what cargo-raze does. It applies the dependency and feature resolution over the root crate, and then reifies that into BUILD files and repository rules. Independent executions of resolution are not composable (and cargo does not do this). What *is* composable though is a set of dependencies (w/ version specifiers) and features. Incidentally, you could consider the lock file as Cargo's product of resolution (though it is incomplete -- it doesn't contain features), and I think we could agree that you can't compose two independently produced lock files.

Because the product of dependency and feature resolution is not composable into a further step of resolution, I assert that `cargo_universe`, which maps to a cargo-raze pass, which maps to one pass of cargo dependency and feature resolution, is not composable with any other `cargo_universe`, and inevitably will not be except in some rare or strange cases, like a library depending on a binary.

--

Now is the time when I review the relevant link on cargo_universe to assess whether my understanding or your understanding was the correct one.

Upon my review I have selected a single piece of evidence to support my claim! That is, that the rule accepts a lock file representing the fully resolved set of dependencies (to be used to bootstrap resolution, same as vanilla cargo). It is *not* a pre resolution artifact (e.g. a workspace, or dependency and feature set), and it will probably not be composable.

**Reply by UebelAndre** [reply]
Lockfiles are generated over cargo workspaces. Not per package. I'm pretty sure I posted some evidence for this in https://github.com/google/cargo-raze/pull/276 but you can also read https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html#creating-a-workspace

So since that PR, cargo raze can now act on the workspace's lockfile which is the resolution of every package. 

Does that clear anything up?

**Reply by acmcarther** [reply]
No, ye be misled!

While the workspace that you compile as part of your application development may have a lock file, (same as any other dependency), your root crate, which is the point of reference during compilation, only considers the lock file of the root crate itself, not of any workspaces it happens to depend upon.

The lock file of a workspace is completely analogous to the lock file of a non-workspace crate. Are you asserting that my_crate which depends on serde consults serde's lockfile during dependency resolution? I am quite confident that it does not (though ready to be challenged). I assert that the same applies in a situation where serde is a crate within a serde-related-workspace.

The lock file only matters for the root crate which is participating in dependency resolution. That root crate can be a workspace root, in which case it is consulted, but this is an uninteresting fact because it is analogous to the non workspace case.

**Reply by UebelAndre** [reply]
> Are you asserting that my_crate which depends on serde consults serde's lockfile during dependency resolution?

No because `serde` is not a cargo workspace memeber. Only workspace members affect the lockfile, right? I'm very confused now. If you look at https://github.com/google/cargo-raze/tree/0026a7bde7487a61aac571ea992fee5cd8a442bd/examples/remote/cargo_workspace

Are you saying that the lockfile there is not what will be used for complication?

**Reply by acmcarther** [reply]
I am asserting one thing. Dependency's lockfiles are not consulted during resolution. If you are cargo-razing (or crate_universing) with against a root crate (or a root workspace), it may possess and use one lock file corresponding to the root toml. Any transitive crates or workspaces you depend on may have a lock file, but these are not used as part of resolution by cargo or cargo-raze, or crate_universe.

A hypothetical crate_universe which solves this problem must not do the dependency resolution -- it must defer it, because it might be used with *other* crate_universes. Similarly, in that situation, it could *not* use a lockfile, because the only lockfile relevant during resolution is the one corresponding to the root toml.

**Reply by UebelAndre** [reply]
Ok, that makes sense. So then the distinction I would make between cargo workspace and a model of `one-list-of-dependencies-to-rule-them-all` is that a cargo workspace may contain multiple versions of the same dependency. But they will all be resolved for the workspace members that depends on them. 

In the `no-internal-toml-model` I would be interested in a way to have a way to model a cargo workspace. In your 3 crates example, each of those crates would be a part of the same universe but that universe can contain multiple versions of the same transitive dependency.

Another good example is again to look at the current state of the examples. I would consider each directory in `examples/[remote|vendored]` to be it's own `crate_universe` as they're already their own cargo workspaces. This makes the `examples/[remote|vendored]/cargo_workspace` example unique in that it contains multiple packages. It's still part of the same cargo workspace, and therefore should be representable in the same `cargo_universe`.

edit: Basically, the lockfiles that are there should be the same lockfiles used in `cargo_universe`, including the `examples/[remote|vendored]/cargo_workspace` lockfiles.

**Reply by acmcarther** [reply]
We're meandering into some additional territory here which is not directly related to the root issue (which is that crates from a given cargo_universe crate may not interact because they each represent an independent execution of crate resolution which might result in transitive deps which are impossible to compile together).

---
> Ok, that makes sense. So then the distinction I would make between cargo workspace and a model of one-list-of-dependencies-to-rule-them-all is that a cargo workspace may contain multiple versions of the same dependency. But they will all be resolved for the workspace members that depends on them.

I interpret this to mean that the individual crates in the workspace can depend upon different versions of the same dependency, which is a capability unique to the workspace feature. As a rule, the aliases (the means by which third party crates are exposed for the no-internal-tomls folks) do not include multiple versions of the same crate.

> In the no-internal-toml-model I would be interested in a way to have a way to model a cargo workspace. In your 3 crates example, each of those crates would be a part of the same universe but that universe can contain multiple versions of the same transitive dependency.

That depends on what you mean by "model a cargo workspace". If all things are allowed, meaning you're not constrained to having a "workspace toml", then you would just have three directories with three BUILD files. If you have some constraints (e.g. you want to retain a workspace toml), then we have to think about a hybrid approach, such as basically what we have now. This needs more thought, but my conclusion so far, at the risk of being tedious, is that you cannot compose two tomls-everywhere-model workspaces just the same as you can't compose two tomls-everywhere crates.

EDIT: this makes zero progress addressing the request of having the different workspace crates use different versions of crates. :shrug: 

> Another good example is again to look at the current state of the examples. I would consider each directory in examples/[remote|vendored] to be it's own crate_universe as they're already their own cargo workspaces. This makes the examples/[remote|vendored]/cargo_workspace example unique in that it contains multiple packages. It's still part of the same cargo workspace, and therefore should be representable in the same cargo_universe.

That's true. The implication here is that none of these directories could ever interoperate. The workspace example has a limited type of interoperation internally (defined by the relationship between the crates in the workspace). Were on the same page that each directory represents one cargo_universe, and that the workspace version gets the privilege of having several crates operate in the context of that universe.

I didn't communicate this elsewhere, but the reason these examples are in independent dirs (and crate_universes) is just so that they can be isolated from one another. Some of the examples involve bringing in weird kinds of dependencies or specifying features, or whatever, and it's nice if they break independently and not all at once.

---

I recognize there's an unanswered question here, which is "ok, so what do we do about the lack of cargo_universe composability". I don't know the answer to that. I think the answer is probably either (1) treat your code as separate from the rest of the universe, meaning no-internal-tomls, because native bazel rules are always composable at least within the bazel workspace, or (2) treat all code, including your own code, as third party code. I think this a strange reframing or evolution of the "tomls-everywhere-model".

Incidentally, I also recognize that (1) also means "make your code not composable with any existing code in the ecosystem" which seems like throwing the baby out with the bathwater.

**Reply by UebelAndre** [reply]
I think this conversation would best be had over at https://github.com/bazelbuild/rules_rust/issues/2 A lot of good questions and comments for that here. Over the holiday I'm hoping to take some time to do a deep dive into `cargo_universe`. Until then though, There are still a few PRs I'd like to get through which will unblock some efforts on `cargo_universe`.

**Comment by acmcarther** [comment]
When I filed this issue I had the impression that there existed an unverbalized mental model of rust-bazel-cargo-raze where developers are writing new code using new Cargo.tomls, even if such crate is never destined to be pushed to crates.io (or even built with Cargo). In this model, nobody would ever hand write a rust_library because every crate would consist of a toml and either be free floating (and unmoored from any common dependencies), or live in a single large cargo workspace and be "moored" by virtue of a shared lock file. 

Is this indeed the goal we're after? If so, I think we should (informally) come up with a "user journey" for this use case, and a set of goals and non-goals. We can then tie some of these new features being implemented to that larger vision.

If "declare-in-cargo-build-in-bazel" is not the desired end-state, then what specifically is the goal? Is it just to make building a single arbitrary existing crate easier with cargo-raze? What role do rust_library and the other regular rules_rust rules play in this model?

**Reply by UebelAndre** [reply]
I'd love to see a "user journey" list! Happy to help build that out.

I think the ultimate goal of `cargo-raze` is to be the gateway to a world where users can author their dependencies in BUILD files and have some Bazel rule version check and fetch all those dependencies. I only want to define my dependencies once and not end up spending my time writing different variants of the same thing in different places.

However, going back to `declare-in-cargo-build-in-bazel`, I think there's actually a lot of value here. I wrote https://github.com/google/cargo-raze/issues/305#issuecomment-749365129 on just that. Right now `cargo-raze` has some "special" behavior for fetching binary dependencies, a concept that isn't really supported by Cargo just yet. This is necessary for things like [rust_wasm_bindgen](https://github.com/bazelbuild/rules_rust/tree/master/wasm_bindgen) but only due to an implementation detail. (I also think this relates to https://github.com/google/cargo-raze/discussions/327#discussioncomment-235773) If you recall https://github.com/google/cargo-raze/issues/218 and it's relation to `wasm-bindgen` (the issue being `cargo-raze` is no longer able to generate the `wasm-bindgen-cli` binary), binary crates do not appear in Cargo metadata but it's still incredibly useful to be able to rely on those binaries and build other rules with them. I think it'd be appropriate for `cargo-raze` to support a `declare-in-cargo-build-in-bazel` model to solve that issue.

To me, [`cargo_universe`](https://github.com/bazelbuild/rules_rust/issues/2#issuecomment-738803545) is the the first big step towards making Bazel more first class as it could enable the source of truth to be build files. To do this, it's currently using `cargo-raze` as a library since there's a lot of work done here to build a dependency graph for Bazel. But I think the role of `cargo-raze` as a binary should focus on `declare-in-cargo-build-in-bazel` and ultimately be the glue for things like `rust_wasm_bindgen`. I'm very interested in writing a repository rule similar to [new_git_repository](https://docs.bazel.build/versions/master/repo/git.html#new_git_repository) that can take an arbitrary crate from an arbitrary repository (not vendored in a monorepo) and generate build targets for it.

**Reply by acmcarther** [reply]
> I think the ultimate goal of cargo-raze is to be the gateway to a world where users can author their dependencies in BUILD files and have some Bazel rule version check and fetch all those dependencies. I only want to define my dependencies once and not end up spending my time writing different variants of the same thing in different places.

Do you want your code for your application to use hand written BUILD files or hand written Cargo.tomls? 

> I think it'd be appropriate for cargo-raze to support a declare-in-cargo-build-in-bazel model to solve that issue.

I've historically been against the "declare-in-cargo-build-in-bazel" model, hence my push back against features that make BUILD files less clear (as they'd drive us toward the opaque BUILD file model).

(I'm open to it now, for some definition of open, but i don't fully understand the ideas that are taking shape)

> To me, cargo_universe is the the first big step towards making Bazel more first class as it could enable the source of truth to be build files

I agree with this, though not for the same reason. `cargo_universe` appears to allow someone to completely ditch the toml files and just declare a `cargo_universe` in //third_party and write regular BUILD files that depend on those third party deps. Sounds great to me.

**Reply by UebelAndre** [reply]
> Do you want your code for your application to use hand written BUILD files or hand written Cargo.tomls?

There are use cases for both. I've wanted a rule that can setup any crate as a remote repository with a generated build file. This is effectively what's being done for `binary_deps` currently, more or less. But for my main/larger projects, I'll want to hand author my build files.



> I agree with this, though not for the same reason. `cargo_universe` appears to allow someone to completely ditch the toml files and just declare a `cargo_universe` in //third_party and write regular BUILD files that depend on those third party deps. Sounds great to me.

I don't understand why you feel `cargo_universe` is any less of an abstraction. I would want to use some kind of `all_crates` macro for that as well so I don't redefine the same thing multiple times. I see this as being quite similar to python's requirement files in that you can `pip_install` a requirements file and have it be represented by a repository name.

```python
pip_install(   # or pip3_import
   name = "my_deps",
   requirements = "//path/to:requirements.txt",
)
```

I use `all_requirements` because I already hand authored the requirements file to contain only the dependencies for the part of the project I want to use it for. I would fully expect `cargo_universe` to do something similar.

**Reply by acmcarther** [reply]
> I don't understand why you feel cargo_universe is any less of an abstraction. I would want to use some kind of all_crates macro for that as well so I don't redefine the same thing multiple times

I'm not sure if this predates our shared understanding, but from recent comments, I think it might be prudent to clear this up.

People in the "no-internal-tomls" world view generally will have a monorepo consisting of a non-trivial number of rust_library and rust_binary floating around (among other non-rust build targets). In general, no individual BUILD target will likely depend on every single dependency, so it would be a rare circumstance where an "all_crates()" is useful.

(i recognize the utility in an environment where crate_universe maps one to one to internal cargo-tomls, etc, etc)

**Reply by UebelAndre** [reply]
Yeah, this predates a shared understanding. I did not have a world in mind where users had a colossal `crate_universe` where a small subset of dependencies were only relevant to a particular target.

**Comment by UebelAndre** [comment]
I think this might be relevant here, but I think this project is quite a unique mix of different components. For the life of this project, I think it's valuable to continue to publish cargo raze to `crates.io` both for use as a library and as a binary. But I also feel like there's huge potential in being able to build it with Bazel and begin to use it in Bazel rules.

**Comment by UebelAndre** [comment]
One other thing of note, I don't think any of the examples are in alignment with the Google model of cargo-raze.

**Reply by acmcarther** [reply]
As author of the initial examples, I interpreted *all* of those initial as following the original model. I omitted the third_party directory because it didn't enhance clarity.

In what sense do they not follow the original model?

**Reply by UebelAndre** [reply]
Shouldn't there be a singular `third_party` directory? And shouldn't there only be 1 `Cargo.toml` file? The fact that there are multiple crates with multiple `Cargo.toml` seems at odds with the Google model.

**Reply by UebelAndre** [reply]
Additionally, this seems "incorrect" since it doesn't follow the Google model
```
├── BUILD
├── Cargo.lock
├── Cargo.toml
├── README.md
├── cargo
│   ├── BUILD.bazel
│   ├── crates.bzl
│   └── remote
│       └── BUILD.bazel
└── src
    └── main.rs
```

Shouldn't the `Cargo.toml` file be inside `./cargo/`?

**Reply by acmcarther** [reply]
They were in the cargo directory until you moved them in https://github.com/google/cargo-raze/pull/191/files?file-filters%5B%5D=.toml

(a thing I did not notice until now, thank you very much)

**Reply by acmcarther** [reply]
Gaze upon the old structure and you shall behold my ancient designs

https://github.com/google/cargo-raze/tree/bdc7076d4e34b84ca5a7bfba23748bb8605e2909/examples/remote/non_cratesio

**Reply by UebelAndre** [reply]
For months I've been making changes that grow support for the `tomls-everywhere-model` because to me it was the biggest benefit of this tool. If the `no-internal-tomls-model` is the end goal (which I'd like to work towards), then I feel `cargo_universe` should be rolled out and a larger design doc should be written up for exactly how that's going to work. But there's still absolutely a desire for `tomls-everywhere-model` which can easily live side by side with the `no-internal-tomls-model`. The examples are not the "Google"/`no-internal-tomls-model` but they are definitely valid use cases.

**Reply by acmcarther** [reply]
I think we're on the same page here now that we have the right vocabulary. There are valid usecases for both the "tomls-everywhere" and the "no-internal-tomls" approach and I think we should support both use cases. I just historically have had my brain in the "no-internal-tomls" mode, and haven't understood the struggles of someone working toward a "tomls-everywhere" workflow.

**Reply by UebelAndre** [reply]
So what happens now? 😅

**Reply by acmcarther** [reply]
Generally I intend to green light the relevant changes, but I'm going to continue to ask annoying questions like "how will these things interoperate", and "how exactly is this useful" despite the utility of the feature being incredibly obvious to you.

**Reply by UebelAndre** [reply]
I'm fine with that. I'm tying harder and harder to keep my PRs small and focused to make reviewing it and explaining it as easy as possible. I think the relevant PR here is an interesting one because it's was the thing that brought to light that we have not been on the same page for months.
